File: cleanup_urls.cpp
/*
 * cleanup_urls - Remove invalid URLs from the RocksDB state database
 *
 * Every stored URL is matched against the same URL regex used by the scraper.
 * Entries that do not match — including those containing \n, \r, or other
 * control/unprintable characters — are deleted in a single WriteBatch, and
 * the updated database is sealed into a fresh checkpoint on exit.
 *
 * Prerequisites (Ubuntu/Debian):
 *   sudo apt-get install librocksdb-dev
 *
 * Build:
 *   g++ -std=c++23 -O2 -Wall cleanup_urls.cpp -lrocksdb -o cleanup_urls
 *
 * Usage:
 *   ./cleanup_urls [--db PATH] [--dry-run]
 *
 *   --db PATH    RocksDB directory  (default: state/url_state_db,
 *                falls back to state/url_state_db_checkpoint)
 *   --dry-run    Report invalid URLs without deleting them
 *   --help, -h   Show this message
 */

#include "progress_bar.hpp"
#include "url_state_manager.hpp"

#include <filesystem>
#include <iostream>
#include <regex>
#include <string>
#include <vector>

namespace fs = std::filesystem;

// Full-string URL validation regex.  std::regex_match() implicitly anchors
// both ends, so the entire stored key must be a valid URL.
//
// The host section is validated structurally:
//   • IPv6 literals in brackets            e.g. [::1], [::ffff:192.0.2.1]
//   • Dotted-quad IPv4                     e.g. 192.168.1.1
//   • DNS hostname (RFC 1123 labels)       e.g. example.com, xn--nxasmq6b.com
//     — Each label starts and ends with an alphanumeric char (or underscore
//       for internal/non-standard hosts); hyphens are allowed in the middle.
//     — Labels may NOT start or end with a hyphen, and the host may NOT
//       start with $, *, !, or any other non-alphanumeric character.
// Optional RFC 3986 userinfo (user[:pass]@) is permitted before the host.
//
// No icase flag: schemes are already normalised to lowercase on storage
// (scraper.cpp::normalizeScheme), and icase + non-ASCII bytes can trigger
// UB in std::regex on some implementations.
static const std::regex RE_VALID_URL{
    // scheme
    R"re(https?://)re"
    // optional userinfo  (user or user:pass, followed by @)
    R"re((?:[a-zA-Z0-9._~!$&'()*+,;=:%-]+@)?)re"
    // host: IPv6 literal | dotted-quad IPv4 | DNS hostname
    R"re((?:\[[0-9a-fA-F:.]+\]|\d{1,3}(?:\.\d{1,3}){3}|[a-zA-Z0-9_](?:[a-zA-Z0-9_-]*[a-zA-Z0-9_])?(?:\.[a-zA-Z0-9_](?:[a-zA-Z0-9_-]*[a-zA-Z0-9_])?)*))re"
    // optional port
    R"re((?::\d{1,5})?)re"
    // optional path, and/or query/fragment (RFC 3986 allows ?query with empty
    // path, e.g. https://youtube.com?v=X); last char must not be punctuation
    R"re((?:(?:/[^\s<>"{}|\\^`\[\]]*)?(?:[?#][^\s<>"{}|\\^`\[\]]*)?[^\s<>"{}|\\^`\[\].,;:!?'"\)])?)re",
    std::regex::optimize};

static bool isValidUrl(const std::string& url) {
    // Guard against stack overflow: GCC's std::regex uses a recursive NFA
    // whose call depth is O(url.size()) for patterns like [A]+[B].  On a
    // typical 8 MB stack this overflows at ~50 k characters, producing
    // SIGSEGV.  Any URL longer than 8 192 bytes is malformed anyway.
    static constexpr std::size_t MAX_URL_LEN = 8192;
    if (url.size() > MAX_URL_LEN) return false;
    if (url.starts_with("http:///") || url.starts_with("https:///")) return false;
    return std::regex_match(url, RE_VALID_URL);
}

static void usage(const char* prog) {
    std::cout <<
        "Usage: " << prog << " [OPTIONS]\n\n"
        "Options:\n"
        "  --db PATH    RocksDB directory (default: state/url_state_db,\n"
        "               falls back to state/url_state_db_checkpoint)\n"
        "  --dry-run    Report invalid URLs without deleting them\n"
        "  --help, -h   Show this message\n";
}

int main(int argc, char* argv[]) {
    std::ios_base::sync_with_stdio(false);

    std::string dbPath = "state/url_state_db";
    bool        dryRun = false;

    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--dry-run") {
            dryRun = true;
        } else if (a == "--db" && i + 1 < argc) {
            dbPath = argv[++i];
        } else if (a == "--help" || a == "-h") {
            usage(argv[0]);
            return 0;
        } else {
            std::cerr << "Unknown option: " << a << '\n';
            usage(argv[0]);
            return 1;
        }
    }

    // Fall back to the committed checkpoint if the live DB is absent.
    if (!fs::exists(fs::path(dbPath) / "CURRENT")) {
        std::string cpPath = dbPath + "_checkpoint";
        if (fs::exists(fs::path(cpPath) / "CURRENT")) {
            dbPath = cpPath;
        } else {
            std::cerr << "Error: RocksDB not found at '" << dbPath << "'\n"
                      << "       (also checked '" << cpPath << "')\n"
                      << "Run the scraper first, or specify --db PATH.\n";
            return 1;
        }
    }

    // Open without auto-verify to skip the full checksum scan on startup.
    UrlStateManager mgr(dbPath, /*autoVerify=*/false);

    // Pre-count for the progress bar — only when stderr is a TTY so that
    // automated pipelines skip the extra O(N) scan and run at full speed.
    std::size_t total = 0;
    if (isatty(STDERR_FILENO))
        mgr.forEachUrl([&](const std::string&) { ++total; });

    ProgressBar pbar(total, "Scanning");
    std::size_t scanned = 0;
    std::vector<std::string> invalid;

    mgr.forEachUrl([&](const std::string& url) {
        if (!isValidUrl(url))
            invalid.push_back(url);
        pbar.update(++scanned);
    });
    pbar.finish(scanned);

    std::cout << "Scanned : " << scanned        << '\n'
              << "Invalid : " << invalid.size() << '\n';

    if (invalid.empty()) {
        std::cout << "Nothing to remove.\n";
        return 0;
    }

    if (dryRun) {
        std::cout << "(dry-run) Would remove " << invalid.size() << " URL(s):\n";
        for (const auto& u : invalid)
            std::cout << "  " << u << '\n';
        return 0;
    }

    mgr.bulkDelete(invalid);
    std::cout << "Removed : " << invalid.size() << " invalid URL(s)\n";
    // UrlStateManager destructor calls seal(), which writes the updated
    // checkpoint so the cleaned state is persisted for the next run.
    return 0;
}

File: cleanup_urls.py
#!/usr/bin/env python3
"""
URL cleanup script - reads all stored URL chunk files, removes any URL that
does not match the URL regex (including those containing \\n, \\r, or other
control/unprintable characters that corrupt the line-separated text format),
and resaves the files.
"""
import json
import re
import sys
import zipfile
import io
from pathlib import Path

from scraper import (
    STATE_DIR,
    STATE_FILE,
    _load_chunk_from_zip,
    _save_chunk_to_zip,
)

# Full-string URL validation regex: same pattern as scraper._RE_URL but
# anchored so the entire stored value must be a valid URL.  This rejects
# any entry that contains whitespace (including \n, \r, \t) or other
# characters that are illegal inside a URL.
_URL_RE = re.compile(
    r'^https?://[^\s<>"{}|\\^`\[\]]+[^\s<>"{}|\\^`\[\].,;:!?\'\")]$',
    re.IGNORECASE,
)


def _is_valid_url(url: str) -> bool:
    """Return True iff *url* matches the URL regex (no control chars allowed)."""
    return bool(_URL_RE.match(url))


def _load_txt_chunk(path: str) -> list[str]:
    """Load URLs from a plain-text file (one URL per line)."""
    with open(path, 'r', encoding='utf-8', errors='replace') as f:
        return [line for line in f.read().splitlines() if line]


def _save_txt_chunk(urls: list[str], path: str) -> None:
    """Save URLs to a plain-text file (one URL per line)."""
    with open(path, 'w', encoding='utf-8') as f:
        f.write('\n'.join(urls))
        if urls:
            f.write('\n')


def _process_chunk(chunk_file: str) -> tuple[int, int]:
    """
    Load *chunk_file*, filter out invalid URLs, and resave it.

    Returns ``(count_before, count_after)``.  If the file does not exist it
    is silently skipped and ``(0, 0)`` is returned.
    """
    path = Path(chunk_file)
    if not path.exists():
        print(f"  Warning: chunk file not found, skipping: {chunk_file}")
        return 0, 0

    if chunk_file.endswith('.zip'):
        urls = _load_chunk_from_zip(chunk_file)
    else:
        urls = _load_txt_chunk(chunk_file)

    before = len(urls)
    valid = [u for u in urls if _is_valid_url(u)]
    after = len(valid)

    if after < before:
        if chunk_file.endswith('.zip'):
            _save_chunk_to_zip(valid, chunk_file)
        else:
            _save_txt_chunk(valid, chunk_file)
        removed = before - after
        print(f"  {chunk_file}: removed {removed} invalid URL(s) ({after} kept)")
    else:
        print(f"  {chunk_file}: all {before} URL(s) valid")

    return before, after


def cleanup_urls() -> None:
    """Read every stored URL chunk, remove invalid entries, and resave."""
    if not STATE_FILE.exists():
        print(f"Error: state file not found: {STATE_FILE}")
        sys.exit(1)

    with open(STATE_FILE, 'r') as f:
        state = json.load(f)

    print("=" * 80)
    print("URL Cleanup")
    print("=" * 80)

    visited_before = visited_after = 0
    to_visit_before = to_visit_after = 0

    visited_chunks = state.get('visited_urls_chunks', [])
    if visited_chunks:
        print(f"\nProcessing visited URLs ({len(visited_chunks)} chunk file(s)):")
        for cf in visited_chunks:
            b, a = _process_chunk(cf)
            visited_before += b
            visited_after += a

    to_visit_chunks = state.get('to_visit_urls_chunks', [])
    if to_visit_chunks:
        print(f"\nProcessing to-visit URLs ({len(to_visit_chunks)} chunk file(s)):")
        for cf in to_visit_chunks:
            b, a = _process_chunk(cf)
            to_visit_before += b
            to_visit_after += a

    total_removed = (visited_before - visited_after) + (to_visit_before - to_visit_after)

    print(f"\n{'=' * 80}")
    print(f"Visited URLs  : {visited_before} -> {visited_after}"
          f" ({visited_before - visited_after} removed)")
    print(f"To-visit URLs : {to_visit_before} -> {to_visit_after}"
          f" ({to_visit_before - to_visit_after} removed)")
    print(f"Total removed : {total_removed}")
    print(f"{'=' * 80}")

    if total_removed > 0:
        state['visited_urls_count'] = visited_after
        state['to_visit_urls_count'] = to_visit_after
        with open(STATE_FILE, 'w') as f:
            json.dump(state, f, indent=2)
        print(f"State file updated: {STATE_FILE}")


if __name__ == "__main__":
    cleanup_urls()
